File: shared/schemas.ts
// shared/schemas.ts
import { z } from "zod";

// ---- Ingest
export const IngestResponse = z.object({
  doc_id: z.string(),
  title: z.string(),
  pages: z.number(),
  chunks: z.number(),
});
export type IngestResponse = z.infer<typeof IngestResponse>;

// ---- Summarize
export const SummarizeRequest = z.object({ doc_id: z.string() });
export type SummarizeRequest = z.infer<typeof SummarizeRequest>;

export const SummarySection = z.object({
  title: z.string(),
  bullets: z.array(z.string()).default([]),
});
export const SummarizeResponse = z.object({
  summary_sections: z.array(SummarySection).default([]),
});
export type SummarizeResponse = z.infer<typeof SummarizeResponse>;

// ---- Chat
export const ChatMessage = z.object({
  role: z.enum(["user", "assistant"]),
  content: z.string(),
});
export const ChatCitation = z.object({ page: z.number(), snippet: z.string() });
export const ChatRequest = z.object({
  doc_id: z.string(),
  messages: z.array(ChatMessage),
});
export const ChatResponse = z.object({
  answer: z.string(),
  citations: z.array(ChatCitation).default([]),
});
export type ChatRequest = z.infer<typeof ChatRequest>;
export type ChatResponse = z.infer<typeof ChatResponse>;

// ---- Quiz
export const QuizOption = z.object({ id: z.string(), text: z.string() });
export const QuizQuestion = z.object({
  id: z.string(),
  stem: z.string(),
  options: z.array(QuizOption),
  multi: z.boolean().default(true),
});
export const QuizGenerateRequest = z.object({
  doc_id: z.string(),
  n_questions: z.number().default(6),
  type: z.literal("checkbox").default("checkbox"),
});
export const QuizGenerateResponse = z.object({
  quiz_id: z.string(),
  questions: z.array(QuizQuestion),
});
export type QuizGenerateRequest = z.infer<typeof QuizGenerateRequest>;
export type QuizGenerateResponse = z.infer<typeof QuizGenerateResponse>;

export const QuizAnswer = z.object({
  question_id: z.string(),
  option_ids: z.array(z.string()),
});
export const QuizGradeRequest = z.object({
  quiz_id: z.string(),
  answers: z.array(QuizAnswer),
});
export const PerQuestionResult = z.object({
  id: z.string(),
  correct_option_ids: z.array(z.string()),
  rationale: z.string(),
});
export const QuizGradeResponse = z.object({
  score: z.number(),
  per_question: z.array(PerQuestionResult),
});
export type QuizGradeRequest = z.infer<typeof QuizGradeRequest>;
export type QuizGradeResponse = z.infer<typeof QuizGradeResponse>;

// ---- Progress
export const ProgressTotals = z.object({
  docs: z.number(),
  quizzes: z.number(),
  avgScore: z.number(),
});
export const ProgressHistoryPoint = z.object({
  date: z.string(),
  score: z.number(),
});
export const ProgressSkill = z.object({
  name: z.string(),
  level: z.number(),
});
export const ProgressGap = z.object({
  topic: z.string(),
  note: z.string(),
});
export const ProgressResponse = z.object({
  totals: ProgressTotals,
  history: z.array(ProgressHistoryPoint),
  skills: z.array(ProgressSkill),
  gaps: z.array(ProgressGap),
});
export type ProgressResponse = z.infer<typeof ProgressResponse>;

// ---- Suggestions
export const SuggestRequest = z.object({
  gaps: z.array(z.string()),
  time_per_day: z.number(),
  horizon_weeks: z.number(),
});
export const PlanItem = z.object({
  week: z.number(),
  focus: z.string(),
  tasks: z.array(z.string()),
});
export const ResourceItem = z.object({
  title: z.string(),
  type: z.enum(["book", "video", "web"]),
  link: z.string().optional(),
});
export const SuggestResponse = z.object({
  plan: z.array(PlanItem),
  resources: z.array(ResourceItem),
});
export type SuggestRequest = z.infer<typeof SuggestRequest>;
export type SuggestResponse = z.infer<typeof SuggestResponse>;

// ---- Brainrot
export const BrainrotStyle = z.enum(["tldr", "memetic", "motivational"]);
export const BrainrotSummaryRequest = z.object({
  doc_id: z.string(),
  style: BrainrotStyle.default("tldr"),
  duration_sec: z.number().default(30),
});
export const BrainrotSummaryResponse = z.object({
  job_id: z.string(),
  summary_text: z.string(),
  sections: z.array(SummarySection),
});

export const BrainrotTTSRequest = z.object({
  job_id: z.string(),
  text: z.string(),
  voice: z.string().default("en_female_1"),
  speed: z.number().default(1.0),
});
export const BrainrotTTSResponse = z.object({
  audio_path: z.string(),
  duration_ms: z.number(),
});

export const BrainrotCaptionsRequest = z.object({
  job_id: z.string(),
  audio_path: z.string(),
  text: z.string(),
  level: z.enum(["line", "word"]).default("line"),
});
export const BrainrotCaptionsResponse = z.object({
  vtt_path: z.string(),
  srt_path: z.string(),
});

export const BrainrotRenderRequest = z.object({
  job_id: z.string(),
  audio_path: z.string(),
  vtt_path: z.string(),
  aspect: z.enum(["9:16", "16:9"]).default("9:16"),
  theme: z.string().default("siraj"),
  include_waveform: z.boolean().default(true),
});
export const BrainrotRenderResponse = z.object({
  video_path: z.string(),
  thumbnail_path: z.string().optional(),
  duration_ms: z.number(),
});

File: shared/schemas.py
from typing import List, Optional, Literal
from pydantic import BaseModel, Field

# ---- Ingest
class IngestResponse(BaseModel):
    doc_id: str
    title: str
    pages: int
    chunks: int

# ---- Summarize
class SummarizeRequest(BaseModel):
    doc_id: str

class SummarySection(BaseModel):
    title: str
    bullets: List[str] = Field(default_factory=list)

class SummarizeResponse(BaseModel):
    summary_sections: List[SummarySection] = Field(default_factory=list)

# ---- Chat
class ChatMessage(BaseModel):
    role: Literal["user", "assistant"]
    content: str

class ChatCitation(BaseModel):
    page: int
    snippet: str

class ChatRequest(BaseModel):
    doc_id: str
    messages: List[ChatMessage]

class ChatResponse(BaseModel):
    answer: str
    citations: List[ChatCitation] = Field(default_factory=list)

# ---- Quiz
class QuizOption(BaseModel):
    id: str
    text: str

class QuizQuestion(BaseModel):
    id: str
    stem: str
    options: List[QuizOption]
    multi: bool = True

class QuizGenerateRequest(BaseModel):
    doc_id: str
    n_questions: int = 6
    type: Literal["checkbox"] = "checkbox"

class QuizGenerateResponse(BaseModel):
    quiz_id: str
    questions: List[QuizQuestion]

class QuizAnswer(BaseModel):
    question_id: str
    option_ids: List[str]

class QuizGradeRequest(BaseModel):
    quiz_id: str
    answers: List[QuizAnswer]

class PerQuestionResult(BaseModel):
    id: str
    correct_option_ids: List[str]
    rationale: str

class QuizGradeResponse(BaseModel):
    score: int
    per_question: List[PerQuestionResult]

# ---- Progress
class ProgressTotals(BaseModel):
    docs: int
    quizzes: int
    avgScore: float

class ProgressHistoryPoint(BaseModel):
    date: str
    score: float

class ProgressSkill(BaseModel):
    name: str
    level: float

class ProgressGap(BaseModel):
    topic: str
    note: str

class ProgressResponse(BaseModel):
    totals: ProgressTotals
    history: List[ProgressHistoryPoint]
    skills: List[ProgressSkill]
    gaps: List[ProgressGap]

# ---- Suggestions
class SuggestRequest(BaseModel):
    gaps: List[str]
    time_per_day: int
    horizon_weeks: int

class PlanItem(BaseModel):
    week: int
    focus: str
    tasks: List[str]

class ResourceItem(BaseModel):
    title: str
    type: Literal["book", "video", "web"]
    link: Optional[str] = None

class SuggestResponse(BaseModel):
    plan: List[PlanItem]
    resources: List[ResourceItem]

# ---- Brainrot
BrainrotStyle = Literal["tldr", "memetic", "motivational"]

class BrainrotSummaryRequest(BaseModel):
    doc_id: str
    style: BrainrotStyle = "tldr"
    duration_sec: int = 30

class BrainrotSummaryResponse(BaseModel):
    job_id: str
    summary_text: str
    sections: List[SummarySection]

class BrainrotTTSRequest(BaseModel):
    job_id: str
    text: str
    voice: str = "en_female_1"
    speed: float = 1.0

class BrainrotTTSResponse(BaseModel):
    audio_path: str
    duration_ms: int

class BrainrotCaptionsRequest(BaseModel):
    job_id: str
    audio_path: str
    text: str
    level: Literal["line", "word"] = "line"

class BrainrotCaptionsResponse(BaseModel):
    vtt_path: str
    srt_path: str

class BrainrotRenderRequest(BaseModel):
    job_id: str
    audio_path: str
    vtt_path: str
    aspect: Literal["9:16", "16:9"] = "9:16"
    theme: str = "siraj"
    include_waveform: bool = True

class BrainrotRenderResponse(BaseModel):
    video_path: str
    thumbnail_path: Optional[str] = None
    duration_ms: int